keras/backend/common/dtypes.py

```diff
@@ -0,0 +1,310 @@
+import functools
+
+from keras import backend
+from keras.api_export import keras_export
+from keras.backend.common.variables import ALLOWED_DTYPES
+
+# Default dtypes corresponding to Python scalars.
+PYTHON_SCALAR_DTYPES = {
+    bool: "bool",
+    int: "int64",
+    float: "float64",
+    # TODO: support complex dtypes
+    # complex: "complex128",
+}
+BIT64_TO_BIT16_DTYPE = {
+    "int32": "int16",
+    "int64": "int16",
+    "uint32": "uint16",
+    "uint64": "uint16",
+    "float32": "float16",
+    "float64": "float16",
+    # TODO: support complex dtypes
+    # "complex64": "complex32",
+    # "complex128": "complex32",
+}
+BIT64_TO_BIT32_DTYPE = {
+    "int64": "int32",
+    "uint64": "uint32",
+    "float64": "float32",
+    # TODO: support complex dtypes
+    # "complex128": "complex64",
+}
+
+
+@functools.lru_cache(maxsize=None)
+def canonicalize_dtype(float_type, dtype):
+    if float_type == "float16":
+        return BIT64_TO_BIT16_DTYPE.get(dtype, dtype)
+    elif float_type == "float32":
+        return BIT64_TO_BIT32_DTYPE.get(dtype, dtype)
+    elif float_type == "float64":
+        return dtype
+    else:
+        raise ValueError(
+            f"Invalid float_type: {float_type}, `float_type` must be one of"
+            f"('float16', 'float32', 'float64')."
+        )
+
+
+def dtype(x, *, canonicalize=False):
+    """Return the dtype object for a tensor or type."""
```

Comment:
The fact that this can take either a tensor or a dtype is a bit confusing as well.

Why isn't `backend.standardize_dtype(x.dtype)` enough?

Reply:
I have changed to utilize standardize_dtype